### PR TITLE
Render publish

### PR DESCRIPTION
**Pull request by Liametc**

### CODE REVIEW

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from collections import defaultdict
+import errno
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py]: Unused?
```suggestion
```

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        """
+        Initialize the class.
+
+        :param name: The node name.
+        :type name: str
+        :param data: The data used to populate this node.
+        :type data: dict
+        :type parent: QtGui.QWidget
+        """
+        super(RenderPublishWidgetItem, self).__init__(parent=None)
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py]: Use passed in `parent`?
```suggestion
        super(RenderPublishWidgetItem, self).__init__(parent=parent)
```

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+        """
+        Populate a combo box with the data provided.
+
+        :param combo:
+        :type combo: QtGui.QComboBox
+        :param data: The data to set on the combo box.
+        :type data: dict
+        """
+        work_paths = data.get("work_paths", [])
+        for index, path in enumerate(work_paths):
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py]: Might not need to enumerate here as `index` is not used until line 63
```suggestion
        for path in work_paths:
```

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+    """
+    Widget that contains all the :class:`RenderPublishWidgetItem` generated by the 
+    current publisher selection.
+    """
+    def __init__(self, parent=None, description_widget=None):
+        """
+        Initialize the widget.
+
+        :param parent: The parent widget
+        :type: QtGio
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py]:
```suggestion
        :type: QtGui.QWidget
```

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+class RenderPublishWidget(QtGui.QWidget):
+    """
+    Widget that contains all the :class:`RenderPublishWidgetItem` generated by the 
+    current publisher selection.
+    """
+    def __init__(self, parent=None, description_widget=None):
+        """
+        Initialize the widget.
+
+        :param parent: The parent widget
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py:96]: Missing docs for description widget

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+        publish_template = item.properties["publish_template"]
+
+        if not work_template or not publish_template:
+            self.logger.debug("Templates do not exist. Is the context set propertly?")
+            return {"accepted": False, "visible": False, "enabled": False, "checked": False}
+
+        fields = work_template.validate_and_get_fields(path)
+        if not fields:
+            self.logger.debug(
+                "'{}': '{}', is not a valid path".format(node_name, path)
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py:312]: Apparently `node_name` is undefined here and at line 317 :eyes:

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+        work_paths_to_publish = sorted(work_paths, reverse=True)
+        publish_fields = {}
+        for work_path in work_paths:
+            work_fields = work_template.get_fields(work_path)
+            self.logger.debug("Checking '{}'".format(work_path))
+            for publish_path in publish_paths:
+                # cache result as this might be slow
+                if publish_path not in publish_fields:
+                    publish_fields[publish_path] = publish_template.get_fields(publish_path)
+                publish_fields = publish_fields[publish_path]
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py]: I think the `publish_fields` cache is getting wiped out?!

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+        if not work_paths:
+            self.logger.debug(
+                "'{}': No generated work files".format(node_name)
+            )
+            return {"accepted": False, "visible": True, "enabled": True, "checked": False}
+
+        publish_paths = self.sgtk.abstract_paths_from_template(publish_template, fields)
+
+        work_paths_to_publish = sorted(work_paths, reverse=True)
+        publish_fields = {}
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py:327]: Could this be made into `self._published_fields = defaultdict(dict)` and cache `publish_template` and `publish_path` together. e.g. at line 331
```python
self._published_fields[publish_template][publish_path] = publish_template.get_fields(publish_path)
```

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+                    work_paths_to_publish.remove(work_path)
+                    break
+
+        if not work_paths_to_publish:
+            self.logger.debug(
+                "'{}': All versions published".format(node_name)
+            )
+            return {"accepted": False, "visible": True, "enabled": False, "checked": False}
+
+        render_paths = self._get_all_render_node_paths()
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py:348]: Apparently `render_paths` is unused? :eyes:

File: hooks/tk-multi-publish2/basic/publish_renders.py [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+        Validates the given item to check that it is ok to publish. Returns a
+        boolean to indicate validity.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        :returns: True if item is valid, False otherwise.
+        """
+        node_name = item.properties["node"].getName()
```
Comment (j0yu) [inline hooks/tk-multi-publish2/basic/publish_renders.py]: Unused `node_name` here... wait, maybe this should at top of `def accepted()` instead?

**Review by j0yu — CHANGES_REQUESTED** [review]
:+1:

File: hooks/tk-multi-publish2/basic/publish_renders.py (hunk 1 above)
Comment (Liametc) [reply hooks/tk-multi-publish2/basic/publish_renders.py]: resolved in next commit

**Review by j0yu — APPROVED** [review]

**Pull request merged**